codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java

```diff
@@ -0,0 +1,167 @@
+        super(preferDirect);
+        if (format == null) {
+            throw new NullPointerException("format");
+        }
+        this.format = format;
+    }
+
+    /**
+     * Returns a format of current compression algorithm.
+     */
+    public final CompressionFormat format() {
+        return format;
+    }
+
+    /**
+     * Returns {@code true} if and only if the end of the compressed stream
+     * has been reached.
+     */
+    public abstract boolean isClosed();
+
+    /**
+     * Close this {@link CompressionEncoder} and so finish the encoding.
+     *
+     * The returned {@link ChannelFuture} will be notified once the
+     * operation completes.
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    public ChannelFuture close() {
+        return close(ctx().newPromise());
+    }
+
+    /**
+     * Close this {@link CompressionEncoder} and so finish the encoding.
+     * The given {@link ChannelFuture} will be notified once the operation
+     * completes and will also be returned.
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    public ChannelFuture close(final ChannelPromise promise) {
+        final ChannelHandlerContext ctx = ctx();
+        EventExecutor executor = ctx.executor();
+        if (executor.inEventLoop()) {
+            return finishEncode(ctx, promise);
+        } else {
+            final ChannelPromise p = ctx.newPromise();
+            executor.execute(new OneTimeTask() {
+                @Override
+                public void run() {
+                    ChannelFuture f = finishEncode(ctx(), p);
+                    f.addListener(new ChannelPromiseNotifier(promise));
+                }
+            });
+            return p;
+        }
+    }
+
+    @Override
+    public void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
+        ChannelFuture f = finishEncode(ctx, ctx.newPromise());
+        f.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture f) throws Exception {
+                ctx.close(promise);
+            }
+        });
+
+        if (!f.isDone()) {
+            // Ensure the channel is closed even if the write operation completes in time.
+            ctx.executor().schedule(new OneTimeTask() {
+                @Override
+                public void run() {
+                    if (!promise.isDone()) {
+                        ctx.close(promise);
+                    }
+                }
+            }, 10, TimeUnit.SECONDS); // FIXME: Magic number
+        }
+    }
+
+    /**
+     * Finishes current encoding.
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    protected abstract ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise);
```

Comment:
Actually the javadoc is not correct as it not throws but just fail the ChannelPromise.